SSL certificate problem: unable to get local issuer certificate

### Environment
install **cargo** in **MSYS2**

### The step caused the error
when I ran the `make` command

### what is the error

> /mingw32/bin/python src/etc/dl-snapshot.py i686-pc-windows-gnu
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: http://curl.haxx.se/docs/sslcerts.html

> curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
Traceback (most recent call last):
  File "src/etc/dl-snapshot.py", line 96, in <module>
    raise Exception("failed to fetch url")
Exception: failed to fetch url
Makefile:91: recipe for target 'target/snapshot/cargo/bin/cargo.exe' failed
make: *** [target/snapshot/cargo/bin/cargo.exe] Error 1

--- a/.travis.install.deps.sh
+++ b/.travis.install.deps.sh
@@ -20,48 +20,56 @@ if [ "${TRAVIS}" = "true" ] && [ "${target}" = "unknown-linux-gnu" ]; then
     sudo apt-get install g++-multilib lib32stdc++6
 fi
 
-url=https://static-rust-lang-org.s3.amazonaws.com/dist/`cat src/rustversion.txt`
+url=https://static.rust-lang.org/dist/`cat src/rustversion.txt`
 
-# Install both 64 and 32 bit libraries. Apparently travis barfs if you try to
-# just install the right ones? This should enable cross compilation in the
-# future anyway.
+# On unix hosts install both 32 and 64-bit libraries, but respect BITS to
+# determine what arch should be used by default. On windows we don't use 32/64
+# libraries, but instead we install msvc as an alternate architecture.
 if [ -z "${windows}" ]; then
-    rm -rf rustc *.tar.gz
-    curl -O $url/rustc-nightly-i686-$target.tar.gz
-    curl -O $url/rustc-nightly-x86_64-$target.tar.gz
-    tar xfz rustc-nightly-i686-$target.tar.gz
-    tar xfz rustc-nightly-x86_64-$target.tar.gz
-
     if [ "${BITS}" = "32" ]; then
-        src=x86_64
-        dst=i686
+        cargo_extra=x86_64-$target
+        cargo_host=i686-$target
     else
-        src=i686
-        dst=x86_64
+        cargo_extra=i686-$target
+        cargo_host=x86_64-$target
     fi
-    cp -r rustc-nightly-$src-$target/rustc/lib/rustlib/$src-$target \
-          rustc-nightly-$dst-$target/rustc/lib/rustlib
-    (cd rustc-nightly-$dst-$target && \
-     find rustc/lib/rustlib/$src-$target/lib -type f | \
-     sed 's/^rustc\//file:/' >> rustc/manifest.in)
-
-    ./rustc-nightly-$dst-$target/install.sh --prefix=rustc
-    rm -rf rustc-nightly-$src-$target
-    rm -rf rustc-nightly-$dst-$target
-    rm -f rustc-nightly-i686-$target.tar.gz
-    rm -f rustc-nightly-x86_64-$target.tar.gz
+    libdir=lib
 else
-    rm -rf rustc *.exe *.tar.gz
-    if [ "${BITS}" = "64" ]; then
-        triple=x86_64-pc-windows-gnu
+    if [ "${BITS}" = "32" ]; then
+        cargo_host=i686-pc-windows-gnu
+    elif [ "${MSVC}" = "1" ]; then
+        cargo_host=x86_64-pc-windows-msvc
     else
-        triple=i686-pc-windows-gnu
+        cargo_host=x86_64-pc-windows-gnu
+        cargo_extra=x86_64-pc-windows-msvc
     fi
-    curl -O $url/rustc-nightly-$triple.tar.gz
-    tar xfz rustc-nightly-$triple.tar.gz
-    ./rustc-nightly-$triple/install.sh --prefix=rustc
-    rm -rf rustc-nightly-$triple
-    rm -f rustc-nightly-$triple.tar.gz
+    libdir=bin
 fi
+
+rm -rf rustc *.tar.gz
+curl -O $url/rustc-nightly-$cargo_host.tar.gz
+tar xfz rustc-nightly-$cargo_host.tar.gz
+
+if [ ! -z "${cargo_extra}" ]; then
+    curl -O $url/rustc-nightly-$cargo_extra.tar.gz
+    tar xfz rustc-nightly-$cargo_extra.tar.gz
+
+    cp -r rustc-nightly-$cargo_extra/rustc/$libdir/rustlib/$cargo_extra \
+          rustc-nightly-$cargo_host/rustc/$libdir/rustlib
+    cp -r rustc-nightly-$cargo_extra/rustc/$libdir/rustlib/$cargo_extra/bin \
+          rustc-nightly-$cargo_host/rustc/$libdir/rustlib/$cargo_host
+    (cd rustc-nightly-$cargo_host && \
+     find rustc/$libdir/rustlib/$cargo_extra -type f | \
+     sed 's/^rustc\//file:/' >> rustc/manifest.in)
+    (cd rustc-nightly-$cargo_host && \
+     find rustc/$libdir/rustlib/$cargo_host/bin -type f | \
+     sed 's/^rustc\//file:/' >> rustc/manifest.in)
+    rm -rf rustc-nightly-$cargo_extra
+    rm -f rustc-nightly-$cargo_extra.tar.gz
+fi
+
+./rustc-nightly-$cargo_host/install.sh --prefix=rustc
+rm -rf rustc-nightly-$cargo_host
+rm -f rustc-nightly-$cargo_host.tar.gz
 
 set +x

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -10,17 +10,20 @@ name = "cargo"
 path = "src/cargo/lib.rs"
 
 [dependencies]
+advapi32-sys = "0.1"
 curl = "0.2"
 docopt = "0.6"
 env_logger = "0.3"
+filetime = "0.1"
 flate2 = "0.2"
 git2 = "0.2"
 git2-curl = "0.2"
 glob = "0.2"
+kernel32-sys = "0.1"
 libc = "0.1"
 libgit2-sys = "0.2"
 log = "0.3"
-num_cpus = "0.1"
+num_cpus = "0.2"
 regex = "0.1"
 registry = { path = "src/registry" }
 rustc-serialize = "0.3"
@@ -31,12 +34,7 @@ threadpool = "0.1"
 time = "0.1"
 toml = "0.1"
 url = "0.2"
-filetime = "0.1"
-
-[target.i686-pc-windows-gnu]
-dependencies = { winapi = "0.1", advapi32-sys = "0.1", kernel32-sys = "0.1" }
-[target.x86_64-pc-windows-gnu]
-dependencies = { winapi = "0.1", advapi32-sys = "0.1", kernel32-sys = "0.1" }
+winapi = "0.1"
 
 [dev-dependencies]
 tempdir = "0.3"

--- a/Makefile.in
+++ b/Makefile.in
@@ -53,6 +53,19 @@ BIN_TARGETS := cargo
 BIN_TARGETS := $(BIN_TARGETS:src/bin/%.rs=%)
 BIN_TARGETS := $(filter-out cargo,$(BIN_TARGETS))
 
+ifdef CFG_MSVC_INCLUDE_PATH
+export INCLUDE := $(CFG_MSVC_INCLUDE_PATH)
+endif
+ifdef CFG_MSVC_LIB_PATH
+export LIB := $(CFG_MSVC_LIB_PATH)
+endif
+ifdef CFG_MSVC_BIN
+export PATH := $(CFG_MSVC_BIN):$(PATH)
+endif
+ifdef CFG_MSVC_WINDOWS_SDK_DIR
+export PATH := $(CFG_MSVC_WINDOWS_SDK_DIR):$(PATH)
+endif
+
 define DIST_TARGET
 ifdef CFG_ENABLE_OPTIMIZE
 TARGET_$(1) = $$(TARGET_ROOT)/$(1)/release

--- a/configure
+++ b/configure
@@ -369,6 +369,52 @@ if [ "$CFG_SRC_DIR" != "$CFG_BUILD_DIR" ]; then
     err "cargo does not currently support an out-of-tree build dir"
 fi
 
+for i in $CFG_TARGET
+do
+    case $i in
+        x86_64-*-msvc)
+            # Use the REG program to figure out where VS is installed
+            # We need to figure out where cl.exe and link.exe are, so we do some
+            # munging and some probing here. We also look for the default
+            # INCLUDE and LIB variables for MSVC so we can set those in the
+            # build system as well.
+            install=$(reg QUERY \
+                       'HKLM\SOFTWARE\Wow6432Node\Microsoft\VisualStudio\12.0' \
+                       -v InstallDir)
+            need_ok "couldn't find visual studio install root"
+            CFG_MSVC_ROOT=$(echo "$install" | grep InstallDir | sed 's/.*REG_SZ[ ]*//')
+            CFG_MSVC_ROOT=$(dirname "$CFG_MSVC_ROOT")
+            CFG_MSVC_ROOT=$(dirname "$CFG_MSVC_ROOT")
+            CFG_MSVC_BIN="${CFG_MSVC_ROOT}/VC/bin/amd64"
+            CFG_MSVC_CL="${CFG_MSVC_BIN}/cl.exe"
+            CFG_MSVC_LIB="${CFG_MSVC_BIN}/lib.exe"
+            CFG_MSVC_LINK="${CFG_MSVC_BIN}/link.exe"
+
+            vcvarsall="${CFG_MSVC_ROOT}/VC/vcvarsall.bat"
+            CFG_MSVC_INCLUDE_PATH=$(cmd /c "\"$vcvarsall\" amd64 && cmd /c echo %INCLUDE%")
+            need_ok "failed to learn about MSVC's INCLUDE"
+            CFG_MSVC_LIB_PATH=$(cmd /c "\"$vcvarsall\" amd64 && cmd /c echo %LIB%")
+            need_ok "failed to learn about MSVC's LIB"
+            CFG_MSVC_WINDOWS_SDK_DIR=$(cmd /c \
+                "\"$vcvarsall\" amd64 && cmd /c echo %WindowsSdkDir%")
+            need_ok "failed to learn about MSVC's WindowsSdkDir"
+            export CFG_MSVC_WINDOWS_SDK_DIR="${CFG_MSVC_WINDOWS_SDK_DIR}bin/x64"
+
+            putvar CFG_MSVC_ROOT
+            putvar CFG_MSVC_BIN
+            putvar CFG_MSVC_CL
+            putvar CFG_MSVC_LIB
+            putvar CFG_MSVC_LINK
+            putvar CFG_MSVC_INCLUDE_PATH
+            putvar CFG_MSVC_LIB_PATH
+            putvar CFG_MSVC_WINDOWS_SDK_DIR
+            ;;
+
+        *)
+            ;;
+    esac
+done
+
 if [ ! -z "$CFG_ENABLE_NIGHTLY" ]; then
     if [ ! -f .cargo/config ]; then
         mkdir -p .cargo

--- a/src/etc/dl-snapshot.py
+++ b/src/etc/dl-snapshot.py
@@ -72,7 +72,7 @@ triple = "%s-%s-%s" % (arch, vendor, target_os)
 hash = snaps[date][platform]
 
 tarball = 'cargo-nightly-' + triple + '.tar.gz'
-url = 'https://static-rust-lang-org.s3.amazonaws.com/cargo-dist/%s/%s' % \
+url = 'https://static.rust-lang.org/cargo-dist/%s/%s' % \
   (date.strip(), tarball)
 dl_path = "target/dl/" + tarball
 dst = "target/snapshot"

